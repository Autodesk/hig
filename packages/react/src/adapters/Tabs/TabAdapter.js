import React from "react";
import { Tabs } from "hig-vanilla";
import HIGAdaper, {
  MountedByHIGParentList,
  MapsPropToMethod,
  MapsEventListener
} from "../HIGAdapter";

function TabAdapter(props) {
  return (
    <HIGAdaper {...props} displayName="Tab" HIGConstructor={Tabs._partials.Tab}>
      {adapterProps => (
        <div>
          <MountedByHIGParentList mounter="addTab" {...adapterProps} />
          <MapsEventListener
            listener="onClick"
            handler={props.onClick}
            {...adapterProps}
          />
          <MapsPropToMethod
            value={props.label}
            setter="setLabel"
            {...adapterProps}
          />
          <MapsPropToMethod value={props.active} {...adapterProps}>
            {(instance, value) =>
              value ? instance.activate() : instance.deactivate()}
          </MapsPropToMethod>
        </div>
      )}
    </HIGAdaper>
  );
}

export default TabAdapter;
