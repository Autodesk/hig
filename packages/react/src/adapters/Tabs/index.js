export { default } from "./TabsAdapter";
export { default as Tab } from "./TabAdapter";
