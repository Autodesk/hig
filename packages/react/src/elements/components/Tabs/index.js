export { default } from "./Tabs";
export { default as Tab } from "./PlaceholderTab";
