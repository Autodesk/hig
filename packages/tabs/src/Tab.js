import React from "react";
import PropTypes from "prop-types";

import TabPresenter from "./presenters/TabPresenter";

/**
 * @typedef {Object} RenderTabPayload
 * @property {string} key
 * @property {boolean} [active]
 * @property {string} [label]
 * @property {Function} [handleClick]
 */

/**
 * @param {RenderTabPayload} props
 * @returns {JSX.Element}
 */
// eslint-disable-next-line react/prop-types
function renderTab({ handleClick, ...otherProps }) {
  return <TabPresenter onClick={handleClick} {...otherProps} />;
}

/**
 * @typedef {Object} TabProps
 * @property {boolean} [active]
 * @property {string} [children]
 * @property {string} [label]
 * @property {Function} [onClick]
 * @property {string} render A render prop allowing for custom tab components to be rendered
 */

/**
 * This component is a facade for interfacing with the `Tabs` component.
 * The logic within the `Tabs` component is strictly separated from the `TabPresenter`.
 *
 * @param {TabProps} props
 * @returns {null}
 */
export default function Tab() {
  return null;
}

Tab.defaultProps = {
  render: renderTab
};

Tab.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.node,
  label: PropTypes.string,
  onClick: PropTypes.func,
  render: PropTypes.func.isRequired
};
