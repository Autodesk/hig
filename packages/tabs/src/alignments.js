export const alignments = Object.freeze({
  LEFT: "left",
  CENTER: "center",
  RIGHT: "right"
});

export const AVAILABLE_ALIGNMENTS = Object.freeze(Object.values(alignments));
