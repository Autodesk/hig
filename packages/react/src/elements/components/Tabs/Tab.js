import React from "react";
import PropTypes from "prop-types";
import TabAdapter from "../../../adapters/Tabs/TabAdapter";

class Tab extends React.Component {
  handleClick = () => {
    this.props.activateTab(this.props.index);
  };

  render() {
    return <TabAdapter onClick={this.handleClick} {...this.props} />;
  }
}

Tab.propTypes = {
  index: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  activateTab: PropTypes.func.isRequired
};

export default Tab;
