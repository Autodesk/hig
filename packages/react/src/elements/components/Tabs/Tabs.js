import React from "react";
import PropTypes from "prop-types";
import TabsAdapter from "../../../adapters/Tabs/TabsAdapter";
import Tab from "./Tab";

class Tabs extends React.Component {
  state = {
    activeTabIndex: this.defaultActiveTabIndex()
  };

  setActiveTab = activeTabIndex => {
    this.props.onTabChange(activeTabIndex);
    this.setState({ activeTabIndex });
  };

  defaultActiveTabIndex() {
    return this.props.activeTabIndex !== undefined
      ? this.props.activeTabIndex
      : 0;
  }

  renderActiveTabIndex() {
    return this.props.activeTabIndex !== undefined
      ? this.props.activeTabIndex
      : this.state.activeTabIndex;
  }

  render() {
    const activeTabIndex = this.renderActiveTabIndex();
    const tabs = React.Children.map(this.props.children, (tab, index) => {
      const { children, ...tabProps } = tab.props;
      return {
        active: index === activeTabIndex,
        content: children,
        tabProps,
        index
      };
    });
    const activeTab = tabs.find(({ index }) => index === activeTabIndex);

    return (
      <div>
        <TabsAdapter>
          {tabs.map(({ active, tabProps, index }) => (
            <Tab
              {...tabProps}
              active={active}
              activateTab={this.setActiveTab}
              key={index}
              index={index}
            />
          ))}
        </TabsAdapter>
        {activeTab ? activeTab.content : null}
      </div>
    );
  }
}

Tabs.propTypes = {
  onTabChange: PropTypes.func,
  children: PropTypes.node
};

Tabs.defaultProps = {
  onTabChange: () => {},
  children: null
};

Tabs.__docgenInfo = {
  props: {
    children: {
      description: "Accepts Tab components"
    },
    onTabChange: {
      description: "Called when user activates a tab"
    }
  }
};

export default Tabs;
