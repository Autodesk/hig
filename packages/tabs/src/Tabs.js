import React, { Component, Children } from "react";
import Fragment from "render-fragment";
import PropTypes from "prop-types";
import { polyfill } from "react-lifecycles-compat";

import { AVAILABLE_ALIGNMENTS, alignments } from "./alignments";
import TabsPresenter from "./presenters/TabsPresenter";
import Tab from "./Tab";

const FIRST_TAB_INDEX = 0;

/**
 * @typedef {import("./Tab").RenderTabPayload} RenderTabPayload
 */

/**
 * @typedef {Object} TabMeta
 * @property {string} key
 * @property {import("./tab").TabProps} props
 */

/**
 * @typedef {Object} TabsProps
 * @property {string} [align]
 * @property {ReactNode} [children]
 */

/**
 * @typedef {Object} TabsState
 * @property {number} activeTabIndex
 */

/**
 * @param {ReactNode} children
 * @returns {TabMeta[]}
 */
function createTabs(children) {
  return Children.toArray(children).reduce((result, child) => {
    const { type, key, props = {} } = child;

    if (type === Tab) {
      result.push({ key, props });
    }

    return result;
  }, []);
}

class Tabs extends Component {
  static propTypes = {
    /**
     * Specify how to justify the tabs within their container
     */
    align: PropTypes.oneOf(AVAILABLE_ALIGNMENTS),
    /**
     * Accepts Tab components
     */
    children: PropTypes.node,
    /**
     * Called when user activates a tab
     */
    onTabChange: PropTypes.func
  };

  /**
   * @param {TabsProps} nextProps
   * @param {TabsState} prevState
   * @returns {TabsState | null}
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    const { children } = nextProps;
    const nextTabs = createTabs(children);
    const nextActiveTabIndex = nextTabs.findIndex(({ props }) => props.active);
    const { activeTabIndex: prevActiveTabIndex } = prevState;

    if (
      // If no active tab was declared
      nextActiveTabIndex < 0 ||
      // If the declared active tab is the same as the current active tab
      nextActiveTabIndex === prevActiveTabIndex
    ) {
      return null;
    }

    return {
      activeTabIndex: nextActiveTabIndex
    };
  }

  static defaultProps = {
    align: alignments.CENTER,
    onTabChange: () => {}
  };

  /** @type {TabsState} */
  state = {
    activeTabIndex: FIRST_TAB_INDEX
  };

  /** @returns {TabMeta[]} */
  getTabs() {
    return createTabs(this.props.children);
  }

  /** @returns {TabMeta|undefined} */
  getActiveTab() {
    const { activeTabIndex } = this.state;

    return this.getTabs()[activeTabIndex];
  }

  /**
   * @param {number} nextActiveTabIndex
   */
  setActiveTab(nextActiveTabIndex) {
    const { activeTabIndex: prevActiveTabIndex } = this.state;
    const { onTabChange } = this.props;

    if (prevActiveTabIndex === nextActiveTabIndex) return;

    onTabChange(nextActiveTabIndex);
    this.setState({ activeTabIndex: nextActiveTabIndex });
  }

  /**
   * @param {number} index
   * @returns {Function}
   */
  createTabClickHandler(index) {
    return () => {
      this.setActiveTab(index);
    };
  }

  /**
   * @param {TabMeta} tab
   * @param {number} index
   * @returns {JSX.Element}
   */
  renderTab = ({ key, props }, index) => {
    const { render, label } = props;
    const { activeTabIndex } = this.state;
    /** @type {RenderTabPayload} */
    const payload = {
      key,
      label,
      active: activeTabIndex === index,
      handleClick: this.createTabClickHandler(index)
    };

    return render(payload);
  };

  /**
   * @returns {JSX.Element}
   */
  renderTabs() {
    return (
      <TabsPresenter align={this.props.align}>
        {this.getTabs().map(this.renderTab)}
      </TabsPresenter>
    );
  }

  /** @returns {ReactNode} */
  renderContent() {
    const activeTab = this.getActiveTab();

    return activeTab ? activeTab.props.children : null;
  }

  render() {
    return (
      <Fragment data-hig="fragment">
        {this.renderTabs()}
        {this.renderContent()}
      </Fragment>
    );
  }
}

export default polyfill(Tabs);
