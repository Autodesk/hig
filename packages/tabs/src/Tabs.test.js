import React from "react";
import renderer from "react-test-renderer";

import Tab from "./Tab";
import Tabs from "./Tabs";
import { alignments } from "./alignments";

describe("tabs/Tabs", () => {
  it("renders tabs", () => {
    const wrapper = (
      <Tabs align={alignments.RIGHT}>
        <Tab label="foo" onClick={() => {}}>
          bar
        </Tab>
        <Tab label="hello" active>
          world
        </Tab>
        <Tab label="boom" render={({ key }) => <button key={key} />}>
          bang
        </Tab>
      </Tabs>
    );
    const tree = renderer.create(wrapper).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
