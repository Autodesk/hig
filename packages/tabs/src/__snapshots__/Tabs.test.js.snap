// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs/Tabs renders tabs 1`] = `
<div
  data-hig="fragment"
>
  <ul
    className="hig__tabs hig__tabs--align-right"
  >
    <li
      className="hig__tabs__tab"
    >
      <span
        className="hig__tabs__tab-label"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      >
        foo
      </span>
    </li>
    <li
      className="hig__tabs__tab hig__tabs__tab--active"
    >
      <span
        className="hig__tabs__tab-label"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      >
        hello
      </span>
    </li>
    <button />
  </ul>
  world
</div>
`;
